docs/advanced.rst

```diff
@@ -2125,3 +2127,42 @@ type is explicitly allowed.
             };
         }
     };
+
+Multiple Inheritance
+====================
+
+pybind11 can create bindings for types that derive from multiple base types
+(aka. *multiple inheritance*). To do so, specify all bases in the template
+arguments of the ``class_`` declaration:
+
+.. code-block:: cpp
+
+    py::class_<MyType, BaseType1, BaseType2, BaseType3>(m, "MyType")
+       ...
+
+The base types can be specified in arbitrary order, and they can even be
+interspersed with alias types and holder types (discussed earlier in this
+document)---pybind11 will automatically find out which is which. The only
+requirement is that the type to be declared is the first template argument.
+
+There are several caveats:
+
+1. When only one base type is specified for a C++ type that actually has
+   multiple bases, pybind11 will assume that it does not participate in
+   multiple inheritance, which can lead to undefined behavior. In such cases,
+   add the tag ``multiple_inheritance``:
+
+    .. code-block:: cpp
+
+        py::class_<MyType, BaseType2>(m, "MyType", py::multiple_inheritance());
```

Comment:
This would be nicer to add this as a template option annotation, rather than going through the process_attribute mechanism, so that it stays with the base classes, i.e.:
```C++
py::class_<MyType, py::multiple_inheritance, BaseType2>(...)
```
That is pretty easy to do: just add a new predicate for the type to the list of predicates at the beginning of `class_` like this:
```C++
    template <typename T> using is_mi_annotation = std::is_same<T, multiple_inheritance>;
```
then update the is_valid_class_option predicate with `|| is_mi_annotation<T>::value`, and finally, add this to the `class_` constructor:
```C++
    if (detail::any_of_t<is_mi, options...>::value) record.multiple_inheritance = true;
```